File: lib/core/security/hash_service.dart
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

/// SHA-256 integrity hashing service.
///
/// Generates a deterministic hash of the raw image bytes so that
/// any subsequent tampering can be detected by re-hashing.
class HashService {
  const HashService();

  /// Returns the SHA-256 hex digest of [data].
  String generateSha256(Uint8List data) {
    final digest = sha256.convert(data);
    return digest.toString();
  }

  /// Verifies that [data] matches the expected [hash].
  bool verifySha256(Uint8List data, String hash) {
    return generateSha256(data) == hash;
  }
}

File: backend/app/services/hash_service.py
import hashlib

def calculate_sha256(data: bytes) -> str:
    """
    Calculates the SHA-256 hash of a byte sequence.
    """
    sha256_hash = hashlib.sha256()
    sha256_hash.update(data)
    return sha256_hash.hexdigest()

def verify_hash(data: bytes, expected_hash: str) -> bool:
    """
    Verifies if the calculated hash matches the expected hash.
    """
    actual_hash = calculate_sha256(data)
    return actual_hash.lower() == expected_hash.lower()
